text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1029 @@
+// We could have a syntaxic sugar to do this (future RFC?)
+trait ToT<T> {
+    type Value = T;
+}
+impl<T, A> ToT<T> for A { }
+
+
+// Example usage
+// Reverse the tuple and provide a tuple with the hashes of the tuple members
+fn reverse_tuple_and_hash<(..T), (..RevT)>(value: (..T)) -> (<(..T) as Rev>::Value, (..<T as ToT<usize>>::Value)),
+where
+    (..T): Rev<Value = (..RevT)>,
+    ..(T: Hash), {
+
+    let rev_t = <(..T) as Rev>::rev(value);
+
+    // Here we use the identifiers of the reversed variadic tuple and variadic tuple type in the iteration
+    let hashes = (for ref rev_t <RevT> in rev_t <RevT> { 
+        let mut s = DefaultHasher::new();
+        <RevT as Hash>::hash(rev_t, &mut s);
+        s.finish()
+    });
+
+    (
+        (for rev_t @in rev_t { rev_t }), 
+        hashes,
+    )
+}
+```
+
+## Declaring and using multiple variadic tuple type with same arity
+
+In C++ multiple parameter packs can be expanded in a single expansion form as long as the packs have the same number of items, but there is no constraint concerning the declaration.
+
+For Rust, using the syntax `(..(T1, T2, ..., Tn))` embeds the constraint that the variadic tuple types `T1`, `T2`, ..., `Tn` have the same arity. This is more consistent than not grouping the declaration (ie: `(..T1), (..T2), ..., (..Tn))`) because the signature using the declaration contains all the information required.
+
+We don't need to look at the implementation or body code to know the required constraint about the variadic tuple type arities.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+C++11 sets a decent precedent with its variadic templates, which can be used to define type-safe variadic functions, among other things. This RFC is comparable to the design of _type parameter packs_ (variadic tuple type) and _function parameter pack_ (variadic tuple).
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+## Dynamic libraries tuple implementation assumption
+
+When using dynamic libraries, client libraries may relies that the host contains code up to a specific tuple arity. So we need to have a 
+way to enforce the compiler to generate all the implementation up to a specific tuple arity. (12 will keep backward comptibility with current `std` impl)
+
+## Variadic tuple destructuration with multiple `..`
+
+A use case involving destructuration of multiple variadic tuple is a split operator for variadic tuples:
+
+```rust
+fn split<(..L), (..R)>(value: (..L, ..R)) -> ((..L), (..R)) {
+    // This involves multiple `..` patterns which is not allowed
+    // and for a reason, even if we have the variadic tuple types to`"guess"
+    // how the split is performed, it is still not explicit
+    let (l @ .., r @ ..) = value;
+    (l, r)
+}
+```
+
+Maybe this kind of use case can be solved by annotating the binding:
+```rust
+fn split<(..L), (..R)>(value: (..L, ..R)) -> ((..L), (..R)) {
+    let (l @ .. : (..L), r @ .. : (..R)) = value;
```

Comment:
wait, I'm confused, wasn't this syntax already included in the proposal itself? (`let (ref mut @ l .. : (..L), ref mut r @ .. : (..R)) = source;`)

Reply:
Yes, but it was a mistake and intended to be only in the unresolved questions section.